### ISSUE TITLE
Problem in unsubscribe of useSubscription react hook

### ISSUE THREAD

**Issue by reza-ebrahimi** [apollographql/apollo-client#7964]
When react component `unmounts`, the `useSubscription` hook doesn't `unsubscribe` stream in server side.

```js
const TestComponent = () => {
  const { loading, error, data } = useSubscription(SUBSCRIPTION);
  
  useEffect(() => {
    return () => {
      // unsubscribe here
    };
  }, []);

  return <div>TestComponent</div>;
};
```

**Comment by zrcni** [comment]
`useSubscription` already cleans up after itself :broom: :wink:
https://github.com/apollographql/apollo-client/blob/5eb624c08ac6bc6ddfa93607793242cebd75f5b1/src/react/hooks/useSubscription.ts#L41

**Comment by reza-ebrahimi** [comment]
@zrcni I know that, But it doesn't `unsubscribe` the stream in server side. Note that everything works fine with `Graphql playgorund` but not working using `useSubscription` hook. Using the hook, stream not stopping or destroying in server side and still continue to stream values.

**Comment by reza-ebrahimi** [comment]
I have create a test project for both client and server, Check it out: https://github.com/reza-ebrahimi/TestSubscrtiptionBug

**Comment by reza-ebrahimi** [comment]
I have created a test project for both client and server, Check it out: https://github.com/reza-ebrahimi/TestSubscrtiptionBug

**Comment by reza-ebrahimi** [comment]
It seems `useSubscription` hook sends `start` request twice, but `stop` is sent only once. See this: https://github.com/async-graphql/async-graphql/issues/478#issuecomment-819148931

Thanks @sunli829

**Comment by ghempton** [comment]
We are seeing this in our application as well with lots of leaked subscriptions when using the `useSubscription` hook. This seems like a pretty fundamental issue, has anyone taken a look?

**Comment by Venryx** [comment]
I was wondering why the "stop" event was not being sent to my server, and it seems this is why!

Seems like an issue, since it could potentially keep dozens of subscriptions open that are not actually used by the client anymore. (eg. if the user is moving between lots of different pages, each with their own sets of subscriptions)

**Comment by Venryx** [comment]
So I've looked into this a bit more, and the path that is taken from the user calling `subscription.unsubscribe` to the actual sending of the `MessageTypes.GQL_STOP` message is kinda complicated. One could argue that the complexity is needed for the level of abstraction sought -- and maybe that is true -- but it certainly makes debugging this issue difficult.

Here is the path that is taken:
1) User calls `useSubscription`: https://github.com/apollographql/apollo-client/blob/a565fd5036b23810f59b49affc69a36cdb434a55/src/react/hooks/useSubscription.ts#L15
2) When the component unmounts, `unsubscribe` is called (on the observable stored in the hook): https://github.com/apollographql/apollo-client/blob/a565fd5036b23810f59b49affc69a36cdb434a55/src/react/hooks/useSubscription.ts#L120
2.1) That `subscription` variable was set here:
https://github.com/apollographql/apollo-client/blob/a565fd5036b23810f59b49affc69a36cdb434a55/src/react/hooks/useSubscription.ts#L89
2.2) That `observable` variable was set here:
https://github.com/apollographql/apollo-client/blob/a565fd5036b23810f59b49affc69a36cdb434a55/src/react/hooks/useSubscription.ts#L28-L39
2.3) That `client.subscribe` function is defined here:
https://github.com/apollographql/apollo-client/blob/a565fd5036b23810f59b49affc69a36cdb434a55/src/core/ApolloClient.ts#L374-L378
2.4) Which calls this function:
https://github.com/apollographql/apollo-client/blob/a565fd5036b23810f59b49affc69a36cdb434a55/src/core/QueryManager.ts#L847
2.5) (which returns this unsubscribe function in the next step)
3) Then this unsubscribe function is called:
https://github.com/apollographql/apollo-client/blob/a565fd5036b23810f59b49affc69a36cdb434a55/src/core/QueryManager.ts#L900
4) Which calls unsubscribe on an observable wrapper, created either here:
https://github.com/apollographql/apollo-client/blob/a565fd5036b23810f59b49affc69a36cdb434a55/src/core/QueryManager.ts#L992-L994
... or here ...
https://github.com/apollographql/apollo-client/blob/a565fd5036b23810f59b49affc69a36cdb434a55/src/core/QueryManager.ts#L992-L994
5) Either way, that observable wrapper appears to then have an unsubscribe method on it, because it inherits it from the base class Observable, which is actually just a reexport of `Observable` from `zen-observable-ts`, which is a reexport of the `Observable` class here:
https://github.com/zenparsing/zen-observable/blob/328fb66a99242900c0fd4a330e9ff66ecfa3a887/src/Observable.js#L198

I unfortunately could not definitively locate the definition of the `unsubscribe` function. But I think it ends up resolving to this `unsubscribe` method in the base `Observable` class:
https://github.com/zenparsing/zen-observable/blob/328fb66a99242900c0fd4a330e9ff66ecfa3a887/src/Observable.js#L182-L187

Which calls `closeSubscription`:
https://github.com/zenparsing/zen-observable/blob/328fb66a99242900c0fd4a330e9ff66ecfa3a887/src/Observable.js#L84
...and `cleanupSubscription`:
https://github.com/zenparsing/zen-observable/blob/328fb66a99242900c0fd4a330e9ff66ecfa3a887/src/Observable.js#L59

The latter of which appears to call this cleanup function here (back up in the `Concat` class):
https://github.com/apollographql/apollo-client/blob/a565fd5036b23810f59b49affc69a36cdb434a55/src/utilities/observables/Concast.ts#L216-L239

If you read through the `cleanup` function above, we might have a piece of the puzzle: as the comment mentions, the `Concat.cleanup` method does not unsubscribe from the underlying observable.

Now the comment makes it sound like it's not the cleanup function's role to do that unsubscribing.

But the issue is that nowhere in the `Concast` class do I see anything that calls `unsubscribe` (other than an error handler, which may never get called). And the class is indeed the one that is calling subscribe on the underlying observables:
https://github.com/apollographql/apollo-client/blob/a565fd5036b23810f59b49affc69a36cdb434a55/src/utilities/observables/Concast.ts#L207-L210

But maybe the base `Observable` class is supposed to be calling `unsubscribe` on the underlying observables? Maybe.

There are some lines in it that call `unsubscribe` (though who knows in what circumstances):
https://github.com/zenparsing/zen-observable/blob/328fb66a99242900c0fd4a330e9ff66ecfa3a887/src/Observable.js#L229
https://github.com/zenparsing/zen-observable/blob/328fb66a99242900c0fd4a330e9ff66ecfa3a887/src/Observable.js#L239
https://github.com/zenparsing/zen-observable/blob/328fb66a99242900c0fd4a330e9ff66ecfa3a887/src/Observable.js#L345
https://github.com/zenparsing/zen-observable/blob/328fb66a99242900c0fd4a330e9ff66ecfa3a887/src/Observable.js#L390-L391

At this point, this is getting too complex for me to follow, so I've given up for now. If someone could continue the investigation, it would be appreciated. (perhaps one of the apollo-client maintainers?)

**Comment by Venryx** [comment]
I was about to read through the test files to check whether there are tests set up (to confirm that `unsubscribe` is being called on subscriptions when they ought to be dropped), but then I opened this: https://github.com/apollographql/apollo-client/blob/a565fd5036b23810f59b49affc69a36cdb434a55/src/core/__tests__/QueryManager/index.ts

And I saw it has 5,881 lines.

And I quietly backed away, and closed that tab.

**Comment by benjamn** [comment]
Based on all the great details @Venryx shared above, I believe I've made some progress on this problem in PR #9718. Please take a look when you have a chance! `npm i @apollo/client@beta`

**Comment by webatom** [comment]
@benjamn Hey, I have the same problem

Upgrading to 3.7.0-beta.1 did not help :(

**Comment by benjamn** [comment]
@Venryx Did PR #9718 make any difference for you? You can run `npm i @apollo/client@beta` to try those changes.

**Comment by Venryx** [comment]
> @Venryx Did PR #9718 make any difference for you? You can run `npm i @apollo/client@beta` to try those changes.

I'm a bit confused tbh, because just now I went to check the dev-tools WebSocket events for my website, and I am not currently seeing the issue I saw earlier, of the "stop" event never being sent.

My guess is that, for my website, the issue came about due to some higher-level issue that ended up getting resolved from the other changes I've made in the last few months. So unfortunately, I can't confirm whether your patch resolves my original issue.

That said, I notice three people upvoted the comment above saying the problem still exists for them. Maybe one of them has a reproduction repository that they can share?

**Comment by felixlucien** [comment]
Having the same issue here: Subscriptions not re-subbing when device is locked :((

**Comment by tetchel** [comment]
for what it's worth since I found this through google - though there are mixed opinions online (eg. https://github.com/apollographql/apollo-client/issues/7964#issuecomment-817801295, [this SO post](https://stackoverflow.com/questions/60515463/how-to-unsubscribe-with-usequery-and-subscribetomore)) about whether or not `useSubscription` cleans up after itself, it was not doing this for me.

In my use-case I was using `subscribeToMore`. I fixed the cleanup problem by calling `subscribeToMore` in `useEffect` and then using its return value as the cleanup function.

```typescript
useEffect(() => {
  const unsub = subscribeToMore({ ... your options })
  return () => unsub()
})
```

I don't know why the docs don't mention this. https://www.apollographql.com/docs/react/data/subscriptions/#subscribing-to-updates-for-a-query

**Comment by MrCox007** [comment]
Same issue, useSubscriptions is not resubbing after device is locked. I use expo with apollo-client v3 and hasura as backend.

**Comment by ktrzos** [comment]
I experience the same problem 🙁. Refreshing React app with subscriptions active doesn't send `stop` event to backend.

**Comment by alessbell** [comment]
> It seems useSubscription hook sends start request twice, but stop is sent only once. See this: https://github.com/async-graphql/async-graphql/issues/478#issuecomment-819148931

As far as I can tell, this would only ever occur when used in combination with React's `StrictMode`, the issue described in https://github.com/apollographql/apollo-client/issues/6037.

@jazithedev @reza-ebrahimi @ghempton @Venryx would anyone be able to either corroborate that `StrictMode` is involved in your cases, or provide a runnable reproduction?

We have a [CodeSandbox](https://codesandbox.io/s/github/apollographql/react-apollo-error-template) and forkable [GitHub repo](https://github.com/apollographql/react-apollo-error-template) with a `subscriptions.tsx` already configured for subscriptions reproductions. Thanks!

**Comment by jmarthernandez** [comment]
@alessbell I'm also seeing this error outside of strict mode. In my case I have added debug logs to make sure the component is unmounted but I still get updates from the unmounted component.

There seems to be a problem with the subscription endpoint(https://uifesi.sse.codesandbox.io/graphql) so I can't try to recreate the scenario in the sandbox.

**Comment by HelaGone** [comment]
Hello. Any updates on this? I have a similar problem with the socket not closing its connection after the user closes or reloads the screen. The problem I'm facing is that those loose connections are piling up in the ApolloServer until the point where ApolloServer struggles to respond requests and fails. Is there a way to handle loose connections from ApolloServer?

**Comment by phryneas** [comment]
The next minor release of Apollo Client will come with a rewrite of `useSubscription`, so the problem might be solved by that - if it's a problem we are actually able to address.

What @HelaGone mentions here with subscriptions staying open after a screen reload seems like it would be outside of our control here - are you talking a full app restart/browser refresh?

**Comment by HelaGone** [comment]
Yes it is on full reload when I see that the socket never closes/completes. However I've seen that many other sites that use web sockets heavily have the same behavior. That's what makes me wonder if those loose connections should be managed by the server instead?

**Comment by phryneas** [comment]
On a full reload, the browser will just stop all JavaScript on the site, so there's not much "disconnecting" logic we can add, I fear.

That said, it's also up to the browser to close open connections, so either your browser has a bug (which is very unlikely), or the connection is closed correctly, but for some reason your server doesn't register that. I'd investigate your server here.

**Comment by phryneas** [comment]
[Version 3.11](https://github.com/apollographql/apollo-client/releases/tag/v3.11.0) has been released with a rewrite of `useSubscription`.

If you find any problems with `useSubscription` after 3.11, please open a new issue. I'm closing this issue since the code that originally caused this issue is not part of the library anymore.
